src/libopensc/card-epass2003.c

```diff
@@ -2445,33 +2456,33 @@ epass2003_gen_key(struct sc_card *card, sc_epass2003_gen_key_data * data)
 	if (!data->modulus)
 		LOG_FUNC_RETURN(card->ctx, SC_ERROR_OUT_OF_MEMORY);
 
-
-	if(len == 256)
+	if(256 == len)
 	{
 		unsigned char tmp[256] = {0};
-		int len1 = 0;
-		int len2 = 0;
+		int xCoordinateLen = 0;
+		int yCoordinateLen = 0;
 		//get x value
-		if(rbuf[0] == 0x58 )
-		{
-			len1 = rbuf[1];
-			memcpy(&tmp[0],&rbuf[2],len1);
+		if(0x58 == rbuf[0]){
+			xCoordinateLen = rbuf[1];
+			memcpy(&tmp[0], &rbuf[2], xCoordinateLen);
+		}
+		else{
+			LOG_FUNC_RETURN(card->ctx, SC_ERROR_OBJECT_NOT_VALID);
 		}
-
 		//get y value
-		if(rbuf[2+len1] == 0x59)
-		{
-			len2 = rbuf[2+len1+1];
-			memcpy(&tmp[len1],&rbuf[2+len1+2],len2);
+		if(0x59 == rbuf[2+xCoordinateLen]){
+			yCoordinateLen = rbuf[2+xCoordinateLen+1];
+			memcpy(&tmp[xCoordinateLen], &rbuf[2+xCoordinateLen+2], yCoordinateLen);
```

Comment:
again, you're reading from untrusted input (the card's response). This means that you need to perform length checking!